feat(NFT): Add option types and add missing memo parameter (#276)
near/NEPs: specs/Standards/NonFungibleToken/Payout.md

# Standard for a Multiple-Recipient-Payout mechanic on NFT Contracts (NEP-199)

> Version `2.0.0`.

Reviewer: feels like we should bump this a major version?

Author: This shouldn't be a breaking change.
```suggestion
Version `2.1.0`.
```

## (NEP-199)

> Version `2.1.0`.

Reviewer 2: Optional parameters are not serialized to JSON, and thus old front-ends will be still compatible with the new contracts if they don't specify optional parameters, also new front-ends will be also compatible with the old contracts as extra JSON fields are just silently ignored. @mikedotexe Am I missing something?

```

## Reference-level explanation

```rust
/// A mapping of NEAR accounts to the amount each should be paid out, in
/// the event of a token-sale. The payout mapping MUST be shorter than the
/// maximum length specified by the financial contract obtaining this
/// payout data. Any mapping of length 10 or less MUST be accepted by
/// financial contracts, so 10 is a safe upper limit.

> #[Derive(Serialize, Deserialize)]

Author: ```suggestion
#[derive(Serialize, Deserialize)]
```

#[serde(crate = "near_sdk::serde")]
pub struct Payout {
  payout: HashMap<AccountId, U128>,
}

pub trait Payouts {
  /// Given a `token_id` and NEAR-denominated balance, return the `Payout`.
  /// struct for the given token. Panic if the length of the payout exceeds
  /// `max_len_payout.`
  fn nft_payout(&self, token_id: String, balance: U128, max_len_payout: Option<u32>) -> Payout;
  /// Given a `token_id` and NEAR-denominated balance, transfer the token
  /// and return the `Payout` struct for the given token. Panic if the
  /// length of the payout exceeds `max_len_payout.`
  fn nft_transfer_payout(
    &mut self,
    receiver_id: AccountId,
    token_id: String,

>     approval_id: u64,

Reviewer: was this supposed to be optional as well? as I see in this implementation:
https://github.com/TENK-DAO/tenk/blame/d65e15c5fdc4b70d6aed0c8181bce33ea0a16e68/contracts/tenk/src/payout.rs#L73

Author: ```suggestion
    approval_id: Option<u64>,
```

#[derive(Serialize, Deserialize)]
#[serde(crate = "near_sdk::serde")]
pub struct Payout {
  pub payout: HashMap<AccountId, U128>,
  payout: HashMap<AccountId, U128>,
}

pub trait Payouts{
pub trait Payouts {
  /// Given a `token_id` and NEAR-denominated balance, return the `Payout`.
  /// struct for the given token. Panic if the length of the payout exceeds
  /// `max_len_payout.`
  fn nft_payout(&self, token_id: String, balance: U128, max_len_payout: u32) -> Payout;
  fn nft_payout(&self, token_id: String, balance: U128, max_len_payout: Option<u32>) -> Payout;
  /// Given a `token_id` and NEAR-denominated balance, transfer the token
  /// and return the `Payout` struct for the given token. Panic if the
  /// length of the payout exceeds `max_len_payout.`

>   #[payable]

Reviewer 2: > Also the `payable` decorator is removed since the market place handles the transfers.

I don't understand this statement. `payable` attribute guards against using function-call-only access keys, and given that `nft_transfer_payout` is a sibling to `nft_transfer` which is guarded, I believe it is a mistake to drop `payable` here.

Author: Ah good point! I think at the time I thought it was because the caller should handle the payout and that the caller would always be another contract, but this isn't true.

>   fn nft_transfer_payout(

Author: ```suggestion
 #[payable]
  fn nft_transfer_payout(
```

    approval_id: Option<u64>,
    memo: Option<String>,
    balance: U128,
    max_len_payout: Option<u32>,
  ) -> Payout {

>     assert_one_yocto();

Reviewer 2: This can be dropped as it does exactly the same as the `payable` attribute.

Author: ```suggestion
```

    receiver_id: AccountId,
    token_id: String,
    approval_id: u64,
    memo: Option<String>,
    balance: U128,
    max_len_payout: Option<u32>,
  ) -> Payout {
    assert_one_yocto();
    let payout = self.nft_payout(token_id, balance);
    self.nft_transfer(receiver_id, token_id, approval_id, memo);
    payout
  }
}
```

## Fallback on error

> In the case where either the `max_len_payout` causes a panic, or a malformed `Payout` is returned, the caller contract should transfer all funds to the token owner.

Reviewer 3: Example scenario background information:
- Benji puts token A for sale for 100 NEAR. 
- token A has a payout object which dictates that Mike should get 20 NEAR and Benji should get 80 NEAR. 
- Willem comes along and purchases the NFT for 100 NEAR on the marketplace.

Logic: 
- Marketplace contract calls `nft_transfer_payout` on the NFT contract which will transfer the NFT to willem and send back a payout object. 
- That payout object as per the current standard was wrapped in a struct. Your new changes propose it should just be a type (just a hashmap). 

In your new proposal: 
- If the NFT contract sends back a struct: 
```
{
  "payout": {
    "mike": "20000000000000000000000000",
    "benji": "80000000000000000000000000",
  }
}
```
This should transfer all the funds (100 NEAR) to the token owner, which is Willem in this case. That means Willem got the NFT for free essentially (which is fine because so sorry Benji, but your NFT was minted on an improper contract that didn't implement the standards correctly). 

- If the NFT contract sends back a payout object: 
```
{
    "mike": "20000000000000000000000000",
    "benji": "80000000000000000000000000",
}
```
Everything would be fine and the marketplace should pay mike 20 NEAR and benji 80 NEAR.

Author: I'll change it to original `token_owner`.  The seller should be the default.

> In the case where either the `max_len_payout` causes a panic, or a malformed `Payout` is returned, the caller contract should transfer all funds to the token owner.

Author: ```suggestion
In the case where either the `max_len_payout` causes a panic, or a malformed `Payout` is returned, the caller contract should transfer all funds to the original token owner selling the token.
```

## Potential pitfalls

> The payout must include all accounts that should receive funds. Thus the token owner must be included.

Reviewer 3: I wouldn't force the token owner in here. What if an NFT contract wanted an NFT where the owner didn't get any funds?

Author: This should be reworded.  The point is that it's a pitfall to assume that you only need to specify those who get royalties and the market will send the rest to the owner.  I agree with the payout being the list of all who should receive funds.

> The payout must include all accounts that should receive funds. Thus the token owner must be included.

Author: ```suggestion
The payout must include all accounts that should receive funds. Thus it is a mistake to assume that the original token owner will receive funds if included.
```

## Errata

> 

Reviewer 3: Can we add this to the errata?

> 

Author: - Version `2.1.0` adds the `memo` parameter, which while used by `nft_transfer` in implementations that follow `2.0.0`, would have encoded used `None`. It also changes `max_len_payout` to an option type.

> 

Author: ```suggestion

- Version `2.1.0` adds a memo parameter to `nft_transfer_payout`, which previously forced implementers of `2.0.0` to pass `None` to the inner `nft_transfer`. Also refactors `max_len_payout` to be an option type.
```

## Errata

- Version `2.1.0` adds a memo parameter to `nft_transfer_payout`, which previously forced implementers of `2.0.0` to pass `None` to the inner `nft_transfer`. Also refactors `max_len_payout` to be an option type.

> Version `2.0.0` contains the intended `approval_id` of `u64` instead of the stringified `U64` version. This was an oversight, but since the standard was live for a few months before noticing, the team thought it best to bump the major version.

Author: ```suggestion
- Version `2.0.0` contains the intended `approval_id` of `u64` instead of the stringified `U64` version. This was an oversight, but since the standard was live for a few months before noticing, the team thought it best to bump the major version.
```
